Validate and throws a custom error when attempting to register invalid hook functions

jhhom opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/fastify/fastify/blob/master/CONTRIBUTING.md

By making a contribution to this project, I certify that:

* (a) The contribution was created in whole or in part by me and I
  have the right to submit it under the open source license
  indicated in the file; or

* (b) The contribution is based upon previous work that, to the best
  of my knowledge, is covered under an appropriate open source
  license and I have the right under that license to submit that
  work with modifications, whether created in whole or in part
  by me, under the same open source license (unless I am
  permitted to submit under a different license), as indicated
  in the file; or

* (c) The contribution was provided directly to me by some other
  person who certified (a), (b) or (c) and I have not modified
  it.

* (d) I understand and agree that this project and the contribution
  are public and that a record of the contribution (including all
  personal information I submit with it, including my sign-off) is
  maintained indefinitely and may be redistributed consistent with
  this project or the open source license(s) involved.
-->

Handles [this issue](https://github.com/fastify/fastify/issues/4320), a hook is invalid if it is not of type function because it needs to bind with the fastify instance.

This PR will add validation for the hook functions passed in the `options` argument of [route registration functions](https://www.fastify.io/docs/latest/Reference/Routes/#shorthand-declaration).

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

cesarvspr reviewed (approved):
lgtm

climba03003 commented on lib/route.js:
```diff
@@ -310,7 +310,12 @@ function buildRouting (options) {
               .concat(opts[hook] || [])
-              .map(h => h.bind(this))
+              .map(h => {
+                if (typeof h !== 'function') {
```
Can you move the logic right after the `onRoute` hook execution.
In here, the `typeof` validation will run through again and again for the same function.

We only need to guarantee the newly added `function` to be valid.

jhhom replied:
Got it, I have moved the logic to right after the `onRoute` hook execution in my latest commit.

Also added new test to pass the coverage check.

climba03003 commented on lib/route.js:
```diff
@@ -243,6 +243,14 @@ function buildRouting (options) {
       }
 
+      for (const hook of lifecycleHooks) {
+        [].concat(opts[hook] || []).forEach(h => {
+          if (typeof h !== 'function') {
+            throw new Error(`${hook} hook should be a function, instead got ${typeof h}`)
+          }
+        })
```
```suggestion
        if(hook in opts) {
          for(const func of opts[hook]) {
            if(typeof func !== 'function') {
              throw new Error(`${hook} hook should be a function, instead got ${typeof func}`)
            }
          }
        }
```
```

Eomm commented on test/hooks.test.js:
```diff
@@ -3301,3 +3301,31 @@ test('onTimeout should be triggered and socket _meta is set', t => {
   })
 })
+
+test('registering invalid hooks should throw an error', async t => {
```
Could you add a test where an `onRoute` hook adds an `undefined` hook?

Here is an example code: https://www.fastify.io/docs/latest/Guides/Plugins-Guide/#hooks

jhhom replied:
Sure, added the test in my latest commit.

delvedor reviewed (changes requested):
Thanks for the pr!
Can you use/create a custom error like we are doing for every other Fastify custom error? See [here](https://github.com/fastify/fastify/blob/main/lib/errors.js).
Thanks!

delvedor reviewed (changes requested):
It seems that we have this error already: https://github.com/fastify/fastify/blob/acba25acc3e97e3598daf79537a39a5272dba7b8/lib/errors.js?rgh-link-date=2022-10-13T09%3A27%3A23Z#L98-L103

mcollina reviewed (approved):
lgtm

jhhom commented:
@delvedor I think the error message I added (e.g "onSend hook should be a function, instead got undefined") can be more informative than the error message in the existing error "The hook callback must be a function", it also provides the following additional information which may helps in debugging:

* which hook is invalid
* the actual type passed

In my latest commit, I modified the error message of the existing error instead.
Just that I'm not sure if modifying the error message of the existing error might have any backwards-compatibility issues. I think its probably not an issue because error checking should be done using the error code.

delvedor commented:
I see your point, then I'll suggest to update the existing error message :)

> Just that I'm not sure if modifying the error message of the existing error might have any backwards-compatibility issues. I think its probably not an issue because error checking should be done using the error code.

This is a valid point, but we have error codes for a reason. Error messages are the human-friendly version of them which is subjected to change.

jhhom commented:
> I see your point, then I'll suggest to update the existing error message :)

the existing error message is updated in my last commit 👍

mcollina reviewed (approved):
lgtm

mcollina commented:
@delvedor PTAL

cesarvspr reviewed (approved):
Nice

delvedor reviewed (approved):
LGTM
